stdlib/LinearAlgebra/src/symmetriceigen.jl

```diff
@@ -5,13 +5,19 @@
 eigencopy_oftype(A::Hermitian, S) = Hermitian(copytrito!(similar(parent(A), S, size(A)), A.data, A.uplo), sym_uplo(A.uplo))
 eigencopy_oftype(A::Symmetric, S) = Symmetric(copytrito!(similar(parent(A), S, size(A)), A.data, A.uplo), sym_uplo(A.uplo))
 
-# Eigensolvers for symmetric and Hermitian matrices
-eigen!(A::RealHermSymComplexHerm{<:BlasReal,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing) =
-    Eigen(sorteig!(LAPACK.syevr!('V', 'A', A.uplo, A.data, 0.0, 0.0, 0, 0, -1.0)..., sortby)...)
+default_eigen_alg(A) = DivideAndConquer()
 
-function eigen(A::RealHermSymComplexHerm; sortby::Union{Function,Nothing}=nothing)
-    S = eigtype(eltype(A))
-    eigen!(eigencopy_oftype(A, S), sortby=sortby)
+# Eigensolvers for symmetric and Hermitian matrices
+function eigen!(A::RealHermSymComplexHerm{<:BlasReal,<:StridedMatrix}, alg::Algorithm = default_eigen_alg(A); sortby::Union{Function,Nothing}=nothing)
+    if alg === DivideAndConquer()
+        Eigen(sorteig!(LAPACK.syevd!('V', A.uplo, A.data)..., sortby)...)
+    elseif alg === QRIteration()
```

Comment:
Isn't better to check for type?
```julia
elseif alg isa QRIteration
```

Reply:
I don't know. One could perhaps check with `@code_typed eigen!(A, QRIteration())` and see if that constant is propagated and the other branches are eliminated. If not, then the type check should help.

Reply:
I think this is fine:

```julia
julia> @code_typed eigen!(A, LinearAlgebra.QRIteration())
CodeInfo(
1 ─ %1 = Base.getfield(A, :uplo)::Char
│   %2 = Base.getfield(A, :data)::Matrix{Float64}
│   %3 = invoke LinearAlgebra.LAPACK.syev!('V'::Char, %1::Char, %2::Matrix{Float64})::Union{Tuple{Vector{Float64}, Matrix{Float64}}, Vector{Float64}}
│   %4 = Core._apply_iterate(Base.iterate, LinearAlgebra.sorteig!, %3, (nothing,))::Tuple{Vector{Float64}, Matrix{Float64}}
│   %5 = Core.getfield(%4, 1)::Vector{Float64}
│   %6 = Core.getfield(%4, 2)::Matrix{Float64}
│   %7 = %new(Eigen{Float64, Float64, Matrix{Float64}, Vector{Float64}}, %5, %6)::Eigen{Float64, Float64, Matrix{Float64}, Vector{Float64}}
└──      return %7
) => Eigen{Float64, Float64, Matrix{Float64}, Vector{Float64}}

julia> @code_typed eigen!(A, LinearAlgebra.RobustRepresentations())
CodeInfo(
1 ─ %1 = Base.getfield(A, :uplo)::Char
│   %2 = Base.getfield(A, :data)::Matrix{Float64}
│   %3 = invoke LinearAlgebra.LAPACK.syevr!('V'::Char, 'A'::Char, %1::Char, %2::Matrix{Float64}, 0.0::Float64, 0.0::Float64, 0::Int64, 0::Int64, -1.0::Float64)::Tuple{Vector{Float64}, Matrix{Float64}}
│   %4 = Core.getfield(%3, 1)::Vector{Float64}
│   %5 = Core.getfield(%3, 2)::Matrix{Float64}
│   %6 = %new(Eigen{Float64, Float64, Matrix{Float64}, Vector{Float64}}, %4, %5)::Eigen{Float64, Float64, Matrix{Float64}, Vector{Float64}}
└──      return %6
) => Eigen{Float64, Float64, Matrix{Float64}, Vector{Float64}}
```

Even the `sorteig!` is compiled away.